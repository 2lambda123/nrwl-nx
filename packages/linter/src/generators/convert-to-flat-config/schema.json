{
  "$schema": "http://json-schema.org/schema",
  "$id": "ConvertToFlatConfig",
  "cli": "nx",
  "description": "Convert an Nx workspace to a Flat ESLint config.",
  "type": "object",
  "properties": {
    "skipFormat": {
      "type": "boolean",
      "description": "Skip formatting files.",
      "default": false,
      "x-priority": "internal"
    }
  },
  "additionalProperties": false,
  "required": []
}
