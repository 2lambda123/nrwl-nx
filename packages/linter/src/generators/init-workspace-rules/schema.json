{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxInitWorkspaceLintRules",
  "cli": "nx",
  "title": "Init Workspace Lint Rules",
  "description": "Initialize Workspace Lint Rules",
  "type": "object",
  "examples": [
    {
      "command": "g @nrwl/linter:init-workspace-rules",
      "description": "Initialize Workspace Lint Rules"
    }
  ],
  "properties": {},
  "required": []
}
