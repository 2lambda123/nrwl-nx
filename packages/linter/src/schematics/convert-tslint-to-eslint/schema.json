{
  "$schema": "http://json-schema.org/schema",
  "id": "convert-tslint-to-eslint",
  "title": "Convert an Nx Angular project from TSLint to ESLint",
  "description": "NOTE: Does not work in the --dry-run mode",
  "examples": [
    {
      "command": "g convert-tslint-to-eslint myapp",
      "description": "Convert the Nx Angular project `myapp` from TSLint to ESLint"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The name of the project to convert.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Which project would you like to convert from TSLint to ESLint?"
    },
    "removeTSLintIfNoMoreTSLintTargets": {
      "type": "boolean",
      "description": "If this conversion leaves no more TSLint usage in the workspace, it will remove TSLint and related dependencies and configuration",
      "default": true,
      "x-prompt": "Would you like to remove TSLint and its related config if there are no TSLint targets remaining after this conversion?"
    }
  },
  "required": ["project"]
}
