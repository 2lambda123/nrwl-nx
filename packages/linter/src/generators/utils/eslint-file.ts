import { joinPathFragments, Tree, updateJson } from '@nx/devkit';
import { Linter } from 'eslint';
import { useFlatConfig } from '../../utils/flat-config';
import {
  addBlockToFlatConfigExport,
  addCompatToFlatConfig,
  generateAst,
  generateFlatOverride,
  generatePluginExtendsElement,
  mapFilePath,
} from './flat-config/ast-utils';
import ts = require('typescript');

export const eslintConfigFileWhitelist = [
  '.eslintrc',
  '.eslintrc.js',
  '.eslintrc.cjs',
  '.eslintrc.yaml',
  '.eslintrc.yml',
  '.eslintrc.json',
  'eslint.config.js',
];

export const baseEsLintConfigFile = '.eslintrc.base.json';

export function findEslintFile(tree: Tree, projectRoot = ''): string | null {
  if (projectRoot === '' && tree.exists(baseEsLintConfigFile)) {
    return baseEsLintConfigFile;
  }
  for (const file of eslintConfigFileWhitelist) {
    if (tree.exists(joinPathFragments(projectRoot, file))) {
      return file;
    }
  }

  return null;
}

export function addOverrideToLintConfig(
  tree: Tree,
  root: string,
  override: Linter.ConfigOverride<Linter.RulesRecord>
) {
  if (useFlatConfig()) {
    const fileName = joinPathFragments(root, 'eslint.config.js');
    const flatOverride = generateFlatOverride(override, root);
    let content = tree.read(fileName, 'utf8');
    // we will be using compat here so we need to make sure it's added
    if (
      !override.env &&
      !override.extends &&
      !override.plugins &&
      !override.parser
    ) {
      content = addCompatToFlatConfig(content);
    }

    tree.write(fileName, addBlockToFlatConfigExport(content, flatOverride));
  } else {
    const fileName = joinPathFragments(root, '.eslintrc.json');
    updateJson(tree, fileName, (json) => {
      json.overrides ?? [];
      json.overrides.push(override);
      return json;
    });
  }
}

export function addExtendsToLintConfig(
  tree: Tree,
  root: string,
  plugin: string
) {
  if (useFlatConfig()) {
    const fileName = joinPathFragments(root, 'eslint.config.js');
    const pluginExtends = generatePluginExtendsElement([plugin]);
    tree.write(
      fileName,
      addBlockToFlatConfigExport(tree.read(fileName, 'utf8'), pluginExtends)
    );
  } else {
    const fileName = joinPathFragments(root, '.eslintrc.json');
    updateJson(tree, fileName, (json) => {
      json.extends = [plugin, ...(json.extends ?? [])];
      return json;
    });
  }
}

export function addIgnoresToLintConfig(
  tree: Tree,
  root: string,
  ignorePatterns: string[]
) {
  if (useFlatConfig()) {
    const fileName = joinPathFragments(root, 'eslint.config.js');
    const block = generateAst<ts.ObjectLiteralExpression>({
      ignores: ignorePatterns.map((path) => mapFilePath(path, root)),
    });
    tree.write(
      fileName,
      addBlockToFlatConfigExport(tree.read(fileName, 'utf8'), block)
    );
  } else {
    const fileName = joinPathFragments(root, '.eslintrc.json');
    updateJson(tree, fileName, (json) => {
      const ignoreSet = new Set([
        ...(json.ignorePatterns ?? []),
        ...ignorePatterns,
      ]);
      json.ignorePatterns = Array.from(ignoreSet);
      return json;
    });
  }
}
