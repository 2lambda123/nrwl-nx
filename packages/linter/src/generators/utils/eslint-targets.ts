import { Tree } from '@nrwl/devkit';
import { forEachExecutorOptions } from '@nrwl/workspace/src/utilities/executor-options-utils';

export function getEslintTargets(tree: Tree) {
  const eslintTargetNames = new Set<string>();
  forEachExecutorOptions(tree, '@nrwl/linter:eslint', (_, __, target) => {
    eslintTargetNames.add(target);
  });
  return eslintTargetNames;
}
