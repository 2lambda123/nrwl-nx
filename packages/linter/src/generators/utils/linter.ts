export const enum Linter {
  EsLint = 'eslint',
  TsLint = 'tslint',
  None = 'none',
}
