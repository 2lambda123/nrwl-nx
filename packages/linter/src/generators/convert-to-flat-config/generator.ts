import {
  formatFiles,
  getProjects,
  ProjectConfiguration,
  readJson,
  Tree,
  updateNxJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { ConvertToFlatConfigGeneratorSchema } from './schema';
import { findEslintFile } from '../utils/eslint-file';
import { convertEslintJsonToFlatConfig } from './converters/json-converter';

export async function convertToFlatConfigGenerator(
  tree: Tree,
  options: ConvertToFlatConfigGeneratorSchema
) {
  const eslintFile = findEslintFile(tree);
  if (!eslintFile) {
    throw new Error('Could not find root eslint file');
  }
  if (!eslintFile.endsWith('.json')) {
    throw new Error(
      'Only json eslint config files are supported for conversion'
    );
  }

  // rename root eslint config to eslint.config.js
  convertRootToFlatConfig(tree);
  // rename and map files
  const projects = getProjects(tree);
  for (const [project, projectConfig] of projects) {
    convertProjectToFlatConfig(tree, project, projectConfig);
  }
  // replace references in nx.json
  updateNxJsonConfig(tree);
  // install missing packages

  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default convertToFlatConfigGenerator;

function convertRootToFlatConfig(tree: Tree) {
  if (tree.exists('.eslintrc.base.json')) {
    convertConfigToFlatConfig(
      tree,
      '',
      '.eslintrc.base.json',
      'eslint.base.config.js'
    );
  }
  if (tree.exists('.eslintrc.json')) {
    convertConfigToFlatConfig(tree, '', '.eslintrc.json', 'eslint.config.js');
  }
}

function convertProjectToFlatConfig(
  tree: Tree,
  project: string,
  projectConfig: ProjectConfiguration
) {
  if (tree.exists(`${projectConfig.root}/.eslintrc.json`)) {
    if (projectConfig.targets) {
      const eslintTargets = Object.keys(projectConfig.targets).filter(
        (t) => projectConfig.targets[t].executor === '@nrwl/linter:eslint'
      );
      for (const target of eslintTargets) {
        projectConfig.targets[target].options = {
          ...projectConfig.targets[target].options,
          eslintConfig: `${projectConfig.root}/eslint.config.js`,
        };
      }
      updateProjectConfiguration(tree, project, projectConfig);
    }

    convertConfigToFlatConfig(
      tree,
      projectConfig.root,
      '.eslintrc.json',
      'eslint.config.js'
    );
  }
}

// update names of eslint files in nx.json
// and remove eslintignore
function updateNxJsonConfig(tree: Tree) {
  if (tree.exists('nx.json')) {
    const content = tree.read('nx.json', 'utf-8');
    const strippedConfig: string = content
      .replace('.eslintrc.json', 'eslint.config.js')
      .replace('.eslintrc.base.json', 'eslint.config.base.js')
      .replace(/".*\.eslintignore.json",?/, '');
    updateNxJson(tree, JSON.parse(strippedConfig));
  }
}

function convertConfigToFlatConfig(
  tree: Tree,
  root: string,
  source: string,
  target: string
) {
  // read original config
  const config = readJson(tree, `${root}/${source}`);

  convertEslintJsonToFlatConfig(tree, root, config, source, target);
}
