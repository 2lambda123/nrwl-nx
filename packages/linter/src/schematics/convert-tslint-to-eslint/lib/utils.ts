import { join, normalize } from '@angular-devkit/core';
import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { addDepsToPackageJson } from '@nrwl/workspace';
import { allFilesInDirInHost } from '@nrwl/workspace/src/utils/ast-utils';
import * as assert from 'assert';

export function updateArrPropAndRemoveDuplication(
  json: Record<string, any>,
  configBeingExtended: Record<string, any>,
  arrPropName: string,
  deleteIfUltimatelyEmpty: boolean
): void {
  json[arrPropName] = json[arrPropName] || [];
  configBeingExtended[arrPropName] = configBeingExtended[arrPropName] || [];
  json[arrPropName] = json[arrPropName].filter(
    (extended: string) => !configBeingExtended[arrPropName].includes(extended)
  );
  json[arrPropName] = Array.from(new Set(json[arrPropName]));
  if (deleteIfUltimatelyEmpty && json[arrPropName].length === 0) {
    delete json[arrPropName];
  }
}

export function updateObjPropAndRemoveDuplication(
  json: Record<string, any>,
  configBeingExtended: Record<string, any>,
  objPropName: string,
  deleteIfUltimatelyEmpty: boolean
): void {
  json[objPropName] = json[objPropName] || {};
  configBeingExtended[objPropName] = configBeingExtended[objPropName] || {};

  for (const [name, val] of Object.entries(json[objPropName])) {
    const valueOfSamePropInExtendedConfig =
      configBeingExtended[objPropName][name];

    try {
      assert.deepStrictEqual(val, valueOfSamePropInExtendedConfig);
      delete json[objPropName][name];
    } catch {}
  }

  if (deleteIfUltimatelyEmpty && Object.keys(json[objPropName]).length === 0) {
    delete json[objPropName];
  }
}

export function ensureESLintPluginsAreInstalled(
  eslintPluginsToBeInstalled: string[]
): Rule {
  return (_host: Tree, context: SchematicContext) => {
    if (!eslintPluginsToBeInstalled?.length) {
      return;
    }

    const additionalDevDependencies = {};

    for (const pluginName of eslintPluginsToBeInstalled) {
      additionalDevDependencies[pluginName] = 'latest';
    }

    context.logger.info(
      '\nINFO: To most closely match your tslint.json, we will ensure the `latest` version of the following eslint plugin(s) are installed:'
    );
    context.logger.info('\n  - ' + eslintPluginsToBeInstalled.join('\n  - '));
    context.logger.info(
      '\nPlease note, you may later wish to pin these to a specific version number in your package.json, rather than leaving it open to `latest`.\n'
    );

    return addDepsToPackageJson({}, additionalDevDependencies);
  };
}

/**
 * Reliably inferring that something is an Angular project is not super straightforward,
 * particularly in the case of libraries, and we have to try a number of different heuristics:
 *
 * - is it configured to use a builder from @angular-devkit (other than the TSLint one itself of course)
 * - it has a karma.conf.js at the root of the project
 * - it has a jest.config.js at the root of the project which contains "jest-preset-angular"
 *
 * NOTE: we cannot do something like "contains a file ending in .module.ts", because NestJS has used so
 * many of the Angular conventions in its file naming.
 */
export function isAngularProject(tree: Tree, projectConfig: any): boolean {
  /**
   * @angular-devkit target heuristic
   */
  const targets = projectConfig.targets || projectConfig.architect;
  for (const [, targetConfig] of Object.entries(targets)) {
    const executor =
      (targetConfig as any).executor || (targetConfig as any).builder;
    if (
      executor &&
      executor.startsWith('@angular-devkit/') &&
      executor !== '@angular-devkit/build-angular:tslint'
    ) {
      return true;
    }
  }
  /**
   * Unit test configuration heuristic
   */
  if (tree.exists(join(normalize(projectConfig.root), 'karma.conf.js'))) {
    return true;
  }
  if (tree.exists(join(normalize(projectConfig.root), 'jest.config.js'))) {
    if (
      tree
        .read(join(normalize(projectConfig.root), 'jest.config.js').toString())
        .includes('jest-preset-angular')
    ) {
      return true;
    }
  }
  return false;
}

export function isE2EProject(projectName: string, projectConfig: any): boolean {
  const targets = projectConfig.targets || projectConfig.architect;
  return projectName.endsWith('-e2e') || !!targets.e2e;
}
