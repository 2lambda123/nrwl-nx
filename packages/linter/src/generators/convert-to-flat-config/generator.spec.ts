import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree, addProjectConfiguration, updateJson } from '@nx/devkit';

import { convertToFlatConfigGenerator } from './generator';
import { ConvertToFlatConfigGeneratorSchema } from './schema';
import { lintProjectGenerator } from '../lint-project/lint-project';
import { Linter } from '../utils/linter';

describe('convert-to-flat-config generator', () => {
  let tree: Tree;
  const options: ConvertToFlatConfigGeneratorSchema = { skipFormat: false };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    addProjectConfiguration(tree, 'test-lib', {
      root: 'libs/test-lib',
      targets: {},
    });
  });

  it('should run successfully', async () => {
    await lintProjectGenerator(tree, {
      skipFormat: false,
      linter: Linter.EsLint,
      eslintFilePatterns: ['**/*.ts'],
      project: 'test-lib',
      setParserOptionsProject: false,
    });
    await convertToFlatConfigGenerator(tree, options);

    expect(tree.exists('eslint.config.js')).toBeTruthy();
    expect(tree.read('eslint.config.js', 'utf-8')).toMatchSnapshot();
    expect(tree.exists('libs/test-lib/eslint.config.js')).toBeTruthy();
    expect(
      tree.read('libs/test-lib/eslint.config.js', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should add global gitignores', async () => {
    await lintProjectGenerator(tree, {
      skipFormat: false,
      linter: Linter.EsLint,
      eslintFilePatterns: ['**/*.ts'],
      project: 'test-lib',
      setParserOptionsProject: false,
    });
    tree.write('.eslintignore', 'ignore/me');
    await convertToFlatConfigGenerator(tree, options);

    expect(tree.read('eslint.config.js', 'utf-8')).toMatchSnapshot();
  });

  it('should add plugins', async () => {
    await lintProjectGenerator(tree, {
      skipFormat: false,
      linter: Linter.EsLint,
      eslintFilePatterns: ['**/*.ts'],
      project: 'test-lib',
      setParserOptionsProject: false,
    });
    updateJson(tree, '.eslintrc.json', (json) => {
      json.plugins = [
        'eslint-plugin-import',
        'single-name',
        '@scope/with-name',
        '@just-scope',
      ];
      return json;
    });
    await convertToFlatConfigGenerator(tree, options);

    expect(tree.read('eslint.config.js', 'utf-8')).toMatchSnapshot();
  });

  it('should add parser', async () => {
    await lintProjectGenerator(tree, {
      skipFormat: false,
      linter: Linter.EsLint,
      eslintFilePatterns: ['**/*.ts'],
      project: 'test-lib',
      setParserOptionsProject: false,
    });
    updateJson(tree, '.eslintrc.json', (json) => {
      json.parser = '@typescript-eslint/parser';
      return json;
    });
    await convertToFlatConfigGenerator(tree, options);

    expect(tree.read('eslint.config.js', 'utf-8')).toMatchSnapshot();
  });
});
