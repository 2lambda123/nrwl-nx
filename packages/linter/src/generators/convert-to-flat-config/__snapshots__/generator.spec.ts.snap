// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-to-flat-config generator should add global gitignores 1`] = `
"const nxEslintPlugin = require("@nx/eslint-plugin");
const { FlatCompat } = require("@eslint/eslintrc");
const eslintrc = new FlatCompat({
    baseDirectory: __dirname
});
export default [
    { ignores: ["**/*"] },
    { ignores: ["ignore/me"] },
    { plugins: {
            "@nx": nxEslintPlugin
        } },
    {
        files: [
            "*.ts",
            "*.tsx",
            "*.js",
            "*.jsx"
        ],
        rules: { @nx/enforce-module-boundaries: [
                "error",
                {
                    enforceBuildableLibDependency: true,
                    allow: [],
                    depConstraints: [{
                            sourceTag: "*",
                            onlyDependOnLibsWithTags: ["*"]
                        }]
                }
            ] }
    },
    {
        files: [
            "*.ts",
            "*.tsx"
        ],
        extends: ["plugin:@nx/typescript"],
        rules: {}
    },
    {
        files: [
            "*.js",
            "*.jsx"
        ],
        extends: ["plugin:@nx/javascript"],
        rules: {}
    }
];
"
`;

exports[`convert-to-flat-config generator should add parser 1`] = `
"const nxEslintPlugin = require("@nx/eslint-plugin");
const typescriptEslintParser = require("@typescript-eslint/parser");
const { FlatCompat } = require("@eslint/eslintrc");
const eslintrc = new FlatCompat({
    baseDirectory: __dirname
});
export default [
    { ignores: ["**/*"] },
    { ignores: ["node_modules"] },
    { plugins: {
            "@nx": nxEslintPlugin
        } },
    {
        parser: typescriptEslintParser
    },
    {
        files: [
            "*.ts",
            "*.tsx",
            "*.js",
            "*.jsx"
        ],
        rules: { @nx/enforce-module-boundaries: [
                "error",
                {
                    enforceBuildableLibDependency: true,
                    allow: [],
                    depConstraints: [{
                            sourceTag: "*",
                            onlyDependOnLibsWithTags: ["*"]
                        }]
                }
            ] }
    },
    {
        files: [
            "*.ts",
            "*.tsx"
        ],
        extends: ["plugin:@nx/typescript"],
        rules: {}
    },
    {
        files: [
            "*.js",
            "*.jsx"
        ],
        extends: ["plugin:@nx/javascript"],
        rules: {}
    }
];
"
`;

exports[`convert-to-flat-config generator should add plugins 1`] = `
"const eslintPluginImport = require("eslint-plugin-import");
const eslintPluginSingleName = require("eslint-plugin-single-name");
const scopeEslintPluginWithName = require("@scope/eslint-plugin-with-name");
const justScopeEslintPlugin = require("@just-scope/eslint-plugin");
const { FlatCompat } = require("@eslint/eslintrc");
const eslintrc = new FlatCompat({
    baseDirectory: __dirname
});
export default [
    { ignores: ["**/*"] },
    { ignores: ["node_modules"] },
    { plugins: {
            "eslint-plugin-import": eslintPluginImport,
            "single-name": eslintPluginSingleName,
            "@scope/with-name": scopeEslintPluginWithName,
            "@just-scope": justScopeEslintPlugin
        } },
    {
        files: [
            "*.ts",
            "*.tsx",
            "*.js",
            "*.jsx"
        ],
        rules: { @nx/enforce-module-boundaries: [
                "error",
                {
                    enforceBuildableLibDependency: true,
                    allow: [],
                    depConstraints: [{
                            sourceTag: "*",
                            onlyDependOnLibsWithTags: ["*"]
                        }]
                }
            ] }
    },
    {
        files: [
            "*.ts",
            "*.tsx"
        ],
        extends: ["plugin:@nx/typescript"],
        rules: {}
    },
    {
        files: [
            "*.js",
            "*.jsx"
        ],
        extends: ["plugin:@nx/javascript"],
        rules: {}
    }
];
"
`;

exports[`convert-to-flat-config generator should add settings 1`] = `
"const nxEslintPlugin = require("@nx/eslint-plugin");
const { FlatCompat } = require("@eslint/eslintrc");
const eslintrc = new FlatCompat({
    baseDirectory: __dirname
});
export default [
    { ignores: ["**/*"] },
    { ignores: ["node_modules"] },
    { plugins: {
            "@nx": nxEslintPlugin
        } },
    { settings: { sharedData: "Hello" } },
    {
        files: [
            "*.ts",
            "*.tsx",
            "*.js",
            "*.jsx"
        ],
        rules: { @nx/enforce-module-boundaries: [
                "error",
                {
                    enforceBuildableLibDependency: true,
                    allow: [],
                    depConstraints: [{
                            sourceTag: "*",
                            onlyDependOnLibsWithTags: ["*"]
                        }]
                }
            ] }
    },
    {
        files: [
            "*.ts",
            "*.tsx"
        ],
        extends: ["plugin:@nx/typescript"],
        rules: {}
    },
    {
        files: [
            "*.js",
            "*.jsx"
        ],
        extends: ["plugin:@nx/javascript"],
        rules: {}
    }
];
"
`;

exports[`convert-to-flat-config generator should run successfully 1`] = `
"const nxEslintPlugin = require("@nx/eslint-plugin");
const { FlatCompat } = require("@eslint/eslintrc");
const eslintrc = new FlatCompat({
    baseDirectory: __dirname
});
export default [
    { ignores: ["**/*"] },
    { ignores: ["node_modules"] },
    { plugins: {
            "@nx": nxEslintPlugin
        } },
    {
        files: [
            "*.ts",
            "*.tsx",
            "*.js",
            "*.jsx"
        ],
        rules: { @nx/enforce-module-boundaries: [
                "error",
                {
                    enforceBuildableLibDependency: true,
                    allow: [],
                    depConstraints: [{
                            sourceTag: "*",
                            onlyDependOnLibsWithTags: ["*"]
                        }]
                }
            ] }
    },
    {
        files: [
            "*.ts",
            "*.tsx"
        ],
        extends: ["plugin:@nx/typescript"],
        rules: {}
    },
    {
        files: [
            "*.js",
            "*.jsx"
        ],
        extends: ["plugin:@nx/javascript"],
        rules: {}
    }
];
"
`;

exports[`convert-to-flat-config generator should run successfully 2`] = `
"const baseConfig = require('../../.eslintrc.json');
const { FlatCompat } = require('@eslint/eslintrc');
const eslintrc = new FlatCompat({
  baseDirectory: __dirname,
});
export default [
  ...baseConfig,
  { ignores: ['!**/*'] },
  {
    files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
    rules: {},
  },
  {
    files: ['*.ts', '*.tsx'],
    rules: {},
  },
  {
    files: ['*.js', '*.jsx'],
    rules: {},
  },
];
"
`;
